フォームリクエスト(Requestバリデーション)を実装(DAIKI)

i-k-i-a-d opened the pull request:
・バリデーションのためにLoginRateRequestクラスを作成し、そこでルートパラメータから値を取得できるように実装。
・LoginRateRequestで受け取ったcourse_idがattendancesテーブル内にあるidしか受け付けないように実装。
・バリデーションルールを詳細に記述するためにAttendancePeriodRuleクラスを作成。
・AttendancePeriodRuleクラス内でパラメータで受け取ったperiodの値を特定の三つの文字以外受け付けないように実装。
・AttendancePeriodRuleクラス内で利用するためにAttendanceモデル内に期間を示す文字列を定数として記述。
・LoginRateメソッド内でバリデーション後の値を受け取れるように実装。
・LoginRateRequestクラスで内でperiodに対するバリデーションにAttendancePeriodRuleクラスに記述したルールが適用となるようにインスタンス化して実装。

i-k-i-a-d committed:
バリデーションのためにLoginRateRequestクラスを作成し、そこでルートパラメータから値を取得できるように実装。
LoginRateRequestで受け取ったcourse_idがattendancesテーブル内にあるidしか受け付けないように実装。
バリデーションルールを詳細に記述するためにAttendancePeriodRuleクラスを作成。
AttendancePeriodRuleクラス内でパラメータで受け取ったperiodの値を特定の三つの文字以外受け付けないように実装。
AttendancePeriodRuleクラス内で利用するためにAttendanceモデル内に期間を示す文字列を定数として記述。
LoginRateメソッド内でバリデーション後の値を受け取れるように実装。
LoginRateRequestクラスで内でperiodに対するバリデーションにAttendancePeriodRuleクラスに記述したルールが適用となるようにインスタンス化して実装。

i-k-i-a-d committed:
Merge branch 'topic/jka-312/dashboard_login_rate' into feature/daiki/jka-312/implement_login_rate_validation

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -95,18 +96,18 @@ public function show(AttendanceShowRequest $request) {
-        if ($period === "week") {
+        if ($request->period === "week") {
             $periodAgo = $endDate->subWeek(1);
-        } elseif ($period === "month") {
+        } elseif ($request->period === "month") {
             $periodAgo = $endDate->subMonth(1);
-        } elseif ($period === "year") {
+        } elseif ($request->period === "year") {
             $periodAgo = $endDate->subYear(1);
         } 
```
ここの比較文字列も定数を参照するようにしましょう！

kou-tech commented on app/Http/Requests/Instructor/LoginRateRequest.php:
```diff
@@ -0,0 +1,46 @@
+     */
+    public function rules()
+    {
+        $validate = [];
+
+        $validate += [
+            'course_id' => [
+                'exists:attendances,course_id','required','integer'
+            ],
+            'period' => ['required', 'string', new AttendancePeriodRule],
+        ];
+
+        return $validate;
```
コードが少し冗長になっていると思います。次のように連想配列形式で指定するだけで良いです！

````php
    public function rules()
    {
        return [
            'course_id' => ['required', 'integer', 'exists:courses,id'],
            'period' => ['required', 'string', new AttendancePeriodRule],
        ];
    }
````

kou-tech replied:
course_idの存在チェックはcoursesテーブルに対して行いましょう。
存在しない講座のログイン率を取得できるのはエラーとしたいですが、受講生がいない講座のログイン率は0という仕様で満たせると思います。

i-k-i-a-d committed:
パラメータで受け取ったperiodを比較する際の文字列をAttendanceモデル内に定義した定数と比較するように実装。
rulesメソッドの中の戻り値の値を単純な連想配列の形に修正。
course_idの存在チェックのバリデーションにcoursesテーブルのidを使うように実装。
